Fix button style of Mini Cart block

dinhtungdu opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->
Fixes #4679

This PR style the mini cart button to match the design. A new component named `QuantityBadge` was added to insolate the style of the badge. On screens smaller than 768px, only the badge with the count is displayed. On bigger screens, cart total, count, and the cart icon (without the counter) are displayed.

Update [October 28th]: Working on this I found two bugs of `formatPrice` function, please see the screenshots below or check two issues: #5018 and #5019.

<!-- Don't forget to update the title with something descriptive. -->
<!-- If your pull request implements a feature flag, make sure you update [this doc](../docs/blocks/features-and-blocks-behind-a-flag.md) -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)
- [ ] I've updated [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) for any feature flags or experimental interfaces implemented in this pull request.

### Screenshots
**Editor:**

![image](https://user-images.githubusercontent.com/5423135/139178205-8743bd5e-2016-40f0-be22-3df09ca26511.png)

**Front-end - full:**

![image](https://user-images.githubusercontent.com/5423135/139177945-d58cf908-0635-4bd3-ab57-554242f18ead.png)

**Front-end - full - after lazy loading:**

![image](https://user-images.githubusercontent.com/5423135/139178702-bce6fe0f-57a0-443b-b0b3-55ce657006a2.png)

**Front-end - empty:**

![image](https://user-images.githubusercontent.com/5423135/139178031-93fed97e-8bb2-448c-98c2-2350d9e11bd9.png)

**Front-end - empty - after lazy loading:**

![image](https://user-images.githubusercontent.com/5423135/139178091-10029de6-16fd-403d-8fa2-be64d54212c4.png)

**Smaller screens (<768px):**

![image](https://user-images.githubusercontent.com/5423135/139178258-6de651bb-2819-41b0-a40c-682a67cdeb40.png)

### Testing
1. Add a columns block.
2. Add Mini Cart Block into a column.
3. See the design changes: the total, number of items, and the cart icon as the screenshots above.
4. Change the browser width to less than `768px`, see the layout change to only cart icon with a counter badge.

### Changelog

> Fix button style of Mini Cart block for small screens.

dinhtungdu committed:
style mini cart button

dinhtungdu committed:
style the button in the editor

dinhtungdu committed:
udpate php button

dinhtungdu commented:
Note that, in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4907/commits/d5301bd41037b2ec055daf8662b7dcbad74048e2 I try to import SVG files so we can share icons between PHP and TSX. Before that, I hard-coded the content of the cart icon in two places: `quantity-badge/index.tsx` and `BlockTypes/MiniCart.php`.

dinhtungdu committed:
Merge branch 'trunk' into fix/4679-mini-cart-button-style

dinhtungdu committed:
try: import svg files

dinhtungdu commented:
Note 2: IMO, this block tends to use inside a column, so I made the button full width. The problem with this approach is when we use the block outside a column, the cart total and count are on the left edge, and the shopping cart icon is pushed to the right.

If the elements of the button are next to each other, we don't need to make the button full width, but according to the design, we have a flex space between the shopping cart icon and the rest.

cc @Aljullu @vivialice

Aljullu commented on assets/js/blocks/cart-checkout/mini-cart/quantity-badge/index.tsx:
```diff
@@ -0,0 +1,14 @@
+import Cart from './cart.svg';
+
+const QuantityBadge = ( { count }: { count: number } ): JSX.Element => (
+	<span className="quantity-badge">
```
If this component will only be used inside the Mini Cart block (I suppose so given that it's inside the `mini-cart` folder), we should prepend `.wc-block-mini-cart__` to the class name.

This also applies to all other class names, they should always be prefixed. That's important because WC core used generic class names like `.price` or `.sale`, and that caused all kind of conflicts with other plugins and themes in the past, so we are trying to make all class names prefixed, now. :slightly_smiling_face: 

There is some more information about how to prefix CSS class names (and other tips about how CSS is used in the WC Blocks repo) in this document:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/docs/contributors/coding-guidelines.md#css

dinhtungdu replied:
I updated the class and style. Because of the prefix, I moved all nested style rules to the first level.

Aljullu replied:
Sounds good, thanks! I think there are still [a couple of classes](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4907/files#diff-3241a379978fe25150781193c3b72b9542b7f2ca751f437d3a37ffd4e569e71fR25-R26) that need to be updated inside `assets/js/blocks/cart-checkout/mini-cart/edit.tsx`.

Aljullu commented on assets/js/blocks/cart-checkout/mini-cart/quantity-badge/cart.svg:
```diff
@@ -0,0 +1,12 @@
+<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Ideally, this SVG file should be stored inside `assets/js/icons/library`. You can see how to add icons there in this [docs page](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/assets/js/icons#adding-icons).

If you do so, I think you will no longer need `@svgr/webpack` beucase you will be able to import the SVG from a React component like this:

```
import { Icon, myIcon } from '@woocommerce/icons';
<Icon srcElement={ myIcon } />
```

Aljullu reviewed (commented):
I didn't take a full review (leaving this for @frontdevde :smile: ), but left a couple of comments about how we use SVG files and CSS with links to the docs. There are a couple of changes that will need to be done regarding this. Let me know if you have any questions!

Aljullu commented:
Ah, sorry. I missed your comments. :sweat_smile: 

> Note that, in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4907/commits/d5301bd41037b2ec055daf8662b7dcbad74048e2 I try to import SVG files so we can share icons between PHP and TSX. Before that, I hard-coded the content of the cart icon in two places: `quantity-badge/index.tsx` and `BlockTypes/MiniCart.php`.

Hmm, yes, good point about sharing the icon between JS and PHP. Do you think it would still be possible to have a wrapper inside `assets/js/icons/library` so that icon can be imported like all the others?

> Note 2: IMO, this block tends to use inside a column, so I made the button full width. The problem with this approach is when we use the block outside a column, the cart total and count are on the left edge, and the shopping cart icon is pushed to the right.

Yes, that's a good question... I would keep it full width for now, as you did. I agree with you, I expect this block will usually be used inside a column (ie: a sidebar) or inside the Navigation block, so being full width will be ok. In the worst case, this can be revisited later.

dinhtungdu commented:
@Aljullu Thanks for the head-up!

> Do you think it would still be possible to have a wrapper inside assets/js/icons/library so that icon can be imported like all the others?

I don't think we can do so because we put the SVG in JS. What I'm doing in the PHP side is get the content of the SVG file and print it.

It's better to follow our convention to put the SVG in the `assets/js/icons/library`. I think we can accept a duplication on the PHP side for now and try to solve that issue if the issue repeats in the future.

dinhtungdu committed:
Revert "try: import svg files"

This reverts commit d5301bd41037b2ec055daf8662b7dcbad74048e2.

dinhtungdu committed:
fix: add cart icon to library, prefix css classes and update style accordingly

frontdevde commented:
@dinhtungdu Could you please add testing instructions to the PR? Thank you!

frontdevde commented on assets/js/blocks/cart-checkout/mini-cart/block.tsx:
```diff
@@ -128,16 +129,25 @@ const MiniCartBlock = ( {
+						cartItemsCount
+					) }
+				</span>
+				<QuantityBadge count={ cartItems.length } />
```
The `cartItemsCount` variable seems more descriptive. Is there a reason for not using it here?
```suggestion
				<QuantityBadge count={ cartItemsCount } />
```

dinhtungdu replied:
No, it's my mistake, thanks for the head-up!

frontdevde commented on assets/js/icons/library/cart-alt.js:
```diff
@@ -0,0 +1,39 @@
+ */
+import { SVG } from 'wordpress-components';
+
+const cartAlt = (
```
Can we find a more descriptive name than `cartAlt` here? Ideally, we'd want the name alone to make it immediately obvious what the icon is used for.

frontdevde commented on assets/js/blocks/cart-checkout/mini-cart/quantity-badge/style.scss:
```diff
@@ -0,0 +1,30 @@
+}
+
+.wc-block-mini-cart__badge {
+	align-items: center;
```
It'd be neat if the property order here could vaguely follow the guidelines in the [Property Ordering section](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/#property-ordering) of the WordPress CSS Coding Standards. No need to be overly pedantic though, just so we have some basic consistent ordering in place across all the styles you're adding in this PR.

frontdevde commented on assets/js/blocks/cart-checkout/mini-cart/edit.tsx:
```diff
@@ -4,27 +4,38 @@
-					),
-					productCount
-				) }
+				<span className="amount">{ formatPrice( productTotal ) }</span>
```
I assume the classes in `edit` should probably be the same as in `block`?
```suggestion
				<span className="wc-block-mini-cart__amount">{ formatPrice( productTotal ) }</span>
```

frontdevde commented on assets/js/blocks/cart-checkout/mini-cart/edit.tsx:
```diff
@@ -4,27 +4,38 @@
-					productCount
-				) }
+				<span className="amount">{ formatPrice( productTotal ) }</span>
+				<span className="count">
```
```suggestion
				<span className="wc-block-mini-cart__count">
```

frontdevde commented on assets/js/blocks/cart-checkout/mini-cart/style.scss:
```diff
@@ -1,3 +1,50 @@
+	padding-left: 0;
+	padding-right: 0;
+	font-weight: 400;
+	width: 100%;
```
> IMO, this block tends to use inside a column, so I made the button full width. The problem with this approach is when we use the block outside a column, the cart total and count are on the left edge, and the shopping cart icon is pushed to the right.

I agree with the reasoning here to make the button full width. That said, maybe we could use a reasonable max-width to prevent the two sides of the button from being too far apart? @vivialice do you have any thoughts on how the button should behave in scenarios where there's lots of available space?
<img width="857" alt="Screenshot 2021-10-08 at 10 41 08" src="https://user-images.githubusercontent.com/1562646/136526111-9e0855e2-1d78-4b5d-a4a6-9612c1d538c1.png">

vivialice replied:
How come we can't have the cart items and pricing to the left of the icon and just leave the white space to the left?

vivialice replied:
Actually, if we are including the badge, do we need to have the item count to the left? I would just show the price if anything.

dinhtungdu replied:
@vivialice From the engineering perspective, it's possible and easier to always display the badge and the total next to the shopping cart icon. I can see that design in someplace:

![image](https://user-images.githubusercontent.com/5423135/137423171-b3a33777-b6d6-4a44-9ba6-986ea96e6801.png)

What is missing to me (if we decide to follow that way) is the design when we display the total including tax.

vivialice replied:
@dinhtungdu Hey thanks for waiting on this - I've mocked up the totals with the alternate labels. I think that's all you need from me on this but let me know if you need any more input from me. I will link to the Figma file in the P2 comment 😄 
<img width="382" alt="Mini Cart Totals" src="https://user-images.githubusercontent.com/9888201/138461925-c840330c-be36-4a3b-a255-f130f6c4d645.png">

frontdevde reviewed (commented):
Just did an initial first pass at testing the changes.

Visually I wasn't sure if the size of the icon in relation to the text is how we want it? When there's little space between the two, the icon does seem quite large in comparison.
<img width="871" alt="Screenshot 2021-10-08 at 10 53 34" src="https://user-images.githubusercontent.com/1562646/136527910-5b196ffc-6acf-4e57-82b0-5f2d08e70ec5.png">

We might also need to address the state where the available space gets too small for the items to display side by side. Options could be to either drop contents below each other or maybe even to collapse to the mobile representation with the quantity badge (if only CSS Container Queries were a thing already to make this trivial 🙂).

<img width="827" alt="Screenshot 2021-10-08 at 10 51 46" src="https://user-images.githubusercontent.com/1562646/136528235-4034a311-9fd9-4e6e-a671-5dee06d67f56.png">

dinhtungdu committed:
Update assets/js/blocks/cart-checkout/mini-cart/block.tsx

Co-authored-by: Michael P. Pfeiffer <frontdevde@users.noreply.github.com>

dinhtungdu committed:
rename mini cart icon

dinhtungdu committed:
Merge branch 'fix/4679-mini-cart-button-style' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/4679-mini-cart-button-style

dinhtungdu committed:
sort css properties

dinhtungdu committed:
Apply suggestions from code review

Co-authored-by: Michael P. Pfeiffer <frontdevde@users.noreply.github.com>

dinhtungdu commented:
@frontdevde Thanks for the detailed review. I fixed all issues except the design-related ones. For the smaller container width, we can follow the approach Storefront is using, which is not perfect IMO (see the screenshot below). But I would rather wait for @vivialice input. Marking this PR as `need design`.

### Storefront Mini Cart in the small container:

![image](https://user-images.githubusercontent.com/5423135/137114059-dd631650-fab2-4c87-b030-0b2bf3811ca1.png)

dinhtungdu committed:
update button style

dinhtungdu committed:
Merge branch 'trunk' into fix/4679-mini-cart-button-style

dinhtungdu committed:
update badge style, support any cart items count

dinhtungdu committed:
avoid using color, inherit from parent

Aljullu commented on assets/js/blocks/cart-checkout/mini-cart/quantity-badge/style.scss:
```diff
@@ -0,0 +1,30 @@
+
+.wc-block-mini-cart__badge {
+	align-items: center;
+	background: $white;
```
If we are hardcoding the `background` color, we also need to hardcode the text `color` property (ie: `color: #000`). Otherwise, this won't look good in a dark theme:

![imatge](https://user-images.githubusercontent.com/3616980/139216026-da23ce34-3b02-4098-ac94-6b07472fd653.png)

You can reproduce that in Storefront going to the Customizer and setting the background color to black and the text and button color to white.

In addition to that, I think we should add `box-shadow: 0 0 0 2px #fff;`, this is [what we did in the Checkout block](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/5d7fd19327f59a3707d8cc62be7df3553efd4443/assets/js/base/components/cart-checkout/order-summary/style.scss#L67) to ensure the badge border was visible in light and dark themes. It's not ideal, but the best way we have right now to make this kind of components to work in light and dark themes.

Ideally, the trolley icon should change its color too based on the font color. But I'm fine leaving it for a follow-up (ie: leaving it until we implement global styles).

Aljullu commented on assets/js/blocks/cart-checkout/mini-cart/quantity-badge/style.scss:
```diff
@@ -0,0 +1,30 @@
+	align-items: center;
+	display: flex;
+	margin-left: 10px;
+	margin-right: 10px;
```
Could we remove the margins here? I know they are applied to the mobile view, but it's unclear to me why they are needed. :thinking:

dinhtungdu replied:
The reason for them and the two below is: in the previous design, we only display the counter on mobile, on desktop we hide it and show the icon only. I forgot to remove it. Thanks for the catch!

Aljullu commented on assets/js/blocks/cart-checkout/mini-cart/style.scss:
```diff
@@ -1,3 +1,42 @@
+
+.wc-block-mini-cart__quantity-badge {
+	margin-left: auto;
+}
```
Then, we can remove this too.

Aljullu commented on assets/js/blocks/cart-checkout/mini-cart/style.scss:
```diff
@@ -1,3 +1,42 @@
+
+	.wc-block-mini-cart__quantity-badge {
+		margin-right: 0;
+	}
```
And this. :slightly_smiling_face:

Aljullu reviewed (commented):
This is working and looking great! I left some comments regarding the CSS, but overall looking good.

> Update [October 28th]: Working on this I found two bugs of `formatPrice` function, please see the screenshots below or check two issues: #5018 and #5019.

Good catch with these issues! I added them to the Mini Cart epic, I think it would be great to fix them before releasing the Mini Cart block.

dinhtungdu committed:
address code review, improve icon color

dinhtungdu commented:
@Aljullu Thanks for the review, I addressed your comments and add some tweaks. Can you take another look at this?

Test results:
Storefront: 
![image](https://user-images.githubusercontent.com/5423135/139269725-896c2df6-eab6-439e-bb7e-0d0d8fdee93d.png)

2021:
![image](https://user-images.githubusercontent.com/5423135/139269916-024c346f-082f-4d60-b745-26f25ceb61d0.png)

TT1 Block:
![image](https://user-images.githubusercontent.com/5423135/139269982-952a06a7-44d6-40d3-aec6-9900c23bb7bc.png)

dinhtungdu committed:
Merge branch 'trunk' into fix/4679-mini-cart-button-style

dinhtungdu commented:
Note that the unit test is failing because we remove the button text, I'm updating the testcases.

dinhtungdu committed:
fix: update unit test to reflect new changes to the button

Aljullu reviewed (approved):
LGTM :ship:
